rmw_fastrtps_shared_cpp/src/rmw_wait.cpp

```diff
@@ -201,6 +201,17 @@ __rmw_wait(
         if (guard_condition.get_trigger_value()) {
           active = true;
           guard_condition.set_trigger_value(false);
+        } else {
+          switch (event->event_type) {
+            // Both a listener and a WaitSet for the same communication status, there are
+            // 2 triggers. Avoid repeated notification, filter it based trigger value.
```

Comment:
Off topic from the original issue we are trying to solve, but if this should be applied to QoS events as well?

Reply:
https://github.com/eProsima/Fast-DDS/issues/3395#issuecomment-1493706345 (that said that pure duplication of the same matched events.) would be also the solution, right? and it can address this behavior in the core before masking the event here.

Reply:
@fujitatomoya 

> Off topic from the original issue we are trying to solve, but if this should be applied to QoS events as well?

The processing logic of events in rmw_fastrtps is the same. So I guess QoS events has the same problem.
But need to further confirm it.   

> https://github.com/eProsima/Fast-DDS/issues/3395#issuecomment-1493706345 (that said that pure duplication of the same matched events.) would be also the solution, right? and it can address this behavior in the core before masking the event here.

I think this way is suitable for repeated event which has been returned from rmw_wait.   
Change here avoid repeated event returned from rmw_wait.